### ISSUE TITLE
cryptography documentation

### ISSUE THREAD

**Issue by igoralmeida** [WhisperSystems/Flock#75]
I was looking for some kind of document explaining your use of cryptography but could not find any, and the wiki seems empty. So, how are you implementing end-to-end encryption?

**Comment by rhodey** [comment]
Hello all,

Checking in to let you know I've started documenting the Flock sync protocol, you can find the most current revision of this document in the "Sync Protocol" section of [this document](https://github.com/rhodey/flock-desktop/blob/master/README.md)

My hope is to finish protocol documentation in the next week, at that time I will move the document from that repo of mine to a wiki on the OWS Flock repo.

**Comment by untitaker** [comment]
Awesome!

**Comment by untitaker** [comment]
@rhodey What I don't understand is why Flock is using a CardDAV/CalDAV server if the end-to-end-encryption prevents the server from understanding the item contents. IMO that's the only reason why you would want to use such terrible protocols: To stay compatible with the rest of the infrastructure, but that compatibility is lost with Flock. In that case you could use any arbitrary protocol.

**Comment by patcon** [comment]
@untitaker See https://github.com/WhisperSystems/Flock/issues/16#issuecomment-49681527 for the rationale :)

**Comment by untitaker** [comment]
I don't think that explains the choice of WebDAV. Yes, those are popular protocols, but Flock isn't compatible at all, it merely uses them as a way to store its encrypted data.

**Comment by patcon** [comment]
Sorry, to summarize, tools like thunderbird know how to talk and sync with webdav servers. They just don't know how to decrypt/encrypt. If you create a cross-platform local proxy app to route the data through (https://github.com/WhisperSystems/Flock/issues/16), then it can do that part and you can use desktop tools on any platform (osx, windows, linux, etc.) with flock data, using only the proxy as a bridge.

**Comment by untitaker** [comment]
I understand that, but I wonder why Flock would go through the trouble of supporting so many different WebDAV servers instead of

  1. Creating its own sync server and protocol
  2. Creating a WebDAV-proxy (for Thunderbird) to that server and protocol.

**Comment by patcon** [comment]
In theory, there's no "supporting so many different WebDAV servers", as they're using the standard protocol. All servers should work, if them implement the spec correctly. In practice, that's less true. Not sure if they knew that when this was started as a spring break of code project

**Comment by rhodey** [comment]
@untitaker building flock on top of WebDAV was a mistake. WebDAV is bloated as a specification and because of the verbosity of XML it is bloated as a protocol on the wire as well.

The cryptography scheme I described in the protocol doc above could easily be adapted to fit other protocols, I'd really like to see someone do this with the [Remote Storage spec](https://remotestorage.io/). Remote Storage has super lean specifications and is uses JSON.

The ideal Flock Sync server would be protocol agnostic and be able to serialize contacts or calendars into XML or JSON representations depending upon the protocols supported by the client (WebDAV, RemoteStorage, etc).

**Comment by untitaker** [comment]
@rhodey Would you adopt a protocol based on Remote Storage for Flock?

**Comment by untitaker** [comment]
The problem I see with the remote storage spec is that basically it seems to be a filesystem with mimetype annotations. Currently there are a few CalDAV clients which are able to sync only a specific timerange, for performance reasons and also because storage on mobile devices is rather limited. With the CalDAV API, an efficient query for events within a given timerange is possible. I don't see an efficient way to do that with the remoteStorage API, at least not in a way that relies *on the client* to generate indices.

**Comment by gusennan** [comment]
@untitaker your last comment about performance, is that because of your thoughts about the necessity of being able to create indexes on the client, or is there another reason?

**Comment by untitaker** [comment]
@gusennan It's about not having the whole collection on the client, but only a specific range that is relevant for the current time. It's only a performance issue for extremely large collections and also exists in Flock AFAIK.

**Comment by gusennan** [comment]
@untitaker which protocol do you think would be better for than remoteStorage or DAV for flock?

**Comment by untitaker** [comment]
@gusennan Note that I'm not affiliated with Flock. I think we'd have to try it out, but remoteStorage is probably a better fit for Flock than DAV.

**Comment by gusennan** [comment]
@untitaker Thank you for the answers.  Neither am I; just have been browsing this codebase and the comments you made caught my attention--was just curious what you thought.

**Comment by untitaker** [comment]
@gusennan To reiterate, the comments I made about performance are quite irrelevant in the context of end-to-end encryption (and I noticed my mistake only in the last paragraph).

**Comment by untitaker** [comment]
@rhodey A question from a crypto newbie who hasn't used Flock: It isn't documented how you get the server password and *EPASSWORD*. Do you actually have two password fields for the user?

Also this:

>If you're using any reports other than the multiget you're doing something weird anyway.

Mobile clients use reports for calendars to only store a relevant timerange of events on the device. This is useful if your calendar collections are truly massive.

**Comment by patcon** [comment]
@untitaker I think you answer is here: https://github.com/WhisperSystems/Flock/issues/75#issuecomment-61890855

on the flock servers, auth password is the result of using pbkdf2 on your password. when self-hosting, you just have two passwords, one for auth and one for encryption.

**Comment by untitaker** [comment]
Ahh I see. So in the end Flock determines the password I have to set for my WebDAV server.

**Comment by patcon** [comment]
When you self-host, they are separate, so you decide both. When using the OWS service, flock simplifies things by deriving the auth password from your own password that they never see :) They handle registration when you use their servers, so they could easily do that

**Comment by untitaker** [comment]
@rhodey Is there any sort of integrity check that verifies the server didn't just delete some of the encrypted items? Was this sort of threat considered when building the threat model? What lead to the final decisions?
